File: coin-change/coin-change.py
class Solution(object):
    def coinChange(self, coins, amount):
        def coin_change(coins,amount,memo={}):
            if amount in memo:
                return memo[amount]
            if amount<0:
                return -1
            if amount==0:
                return 0
            min_=float("inf")
            for coin in coins:
                res=coin_change(coins,amount-coin,memo)
                if res>=0:
                    min_=min(min_,1+res)
            memo[amount]=min_ if min_!=float("inf") else -1   
            return memo[amount]
        return coin_change(coins,amount)
            
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        

File: coin-change/coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        return dp(coins,amount,new HashMap<>());
        
    }
     public int dp(int[] coins, int amount, Map<Integer, Integer> cache) {
        if (cache.containsKey(amount)) {
            return cache.get(amount);
        }
        if (amount == 0) {
            return 0;
        }
        if (amount < 0) {
            return -1;
        }
        int result = Integer.MAX_VALUE;
        for (int coin : coins) {
            int minCalls =  dp(coins, amount - coin, cache);
            if (minCalls >= 0) {
                result = Math.min(result,1+ minCalls);
            }
        }
        if (result != Integer.MAX_VALUE) {
            cache.put(amount, result);
        } else {
            cache.put(amount, -1);
        }
        return cache.get(amount);
    }
}